### ISSUE TITLE
layout bug

### ISSUE THREAD

**Issue by Klortho** [Klortho/d3-flextree#1]
Discovered a bug in the layout algorithm -- see [this issue](https://github.com/cwi-swat/non-layered-tidy-trees/issues/2). I added test trees grandkids1.json and grandkids2.json to illustrate it.

**Comment by Klortho** [comment]
See [this patch](https://gist.github.com/lianyi/250f3db09e17ba0f9020) from @lianyi. It seems to fix this bug, but causes test19 to fail.

 I'm working in

* d3-flextree, layout-bug-grandkids branch,
* d3, flextree-layout-bug branch

**Comment by Klortho** [comment]
The differences for test19 were very small, and seemed, possibly, to be due to the same bug, but on a node deeper in the tree. I'm not 100% sure, but going cross-eyed looking at it.  So, I pulled the change: https://github.com/Klortho/d3/pull/1.

**Comment by Klortho** [comment]
The bug is fixed for this tree:

```javascript
[ 40, 40,
  [ 40, 40 ],
  [ 40, 40,
    [ 100, 40 ],
    [ 200, 40 ],
  ],
],
```

Here's before and after:

<img width="313" alt="tree-a-before" src="https://user-images.githubusercontent.com/77226/36885101-dd638d40-1db2-11e8-8a6b-a407aeafa2a1.png">

<img width="311" alt="tree-a-after" src="https://user-images.githubusercontent.com/77226/36885128-11a2c9cc-1db3-11e8-867a-72642808a80d.png">

But still exists for this tree:

```javascript
[ 40, 40,
  [ 40, 40 ],
  [ 40, 40,
    [ 40, 40,
      [ 100, 40 ],
      [ 200, 40 ],
    ],
  ],
],
```

<img width="312" alt="tree-b-before" src="https://user-images.githubusercontent.com/77226/36885123-095ba914-1db3-11e8-9e2f-aabb3f9f7dcf.png">

**Comment by jiakuan** [comment]
I saw the same bug as well. The algorithm only moves subtree from the second index, that means the first child will never be moved if it doesn't have children.

As an improvement to the algorithm, perhaps we could add a special check for each first leaf child and move the first child instead?

**Comment by Klortho** [comment]
This is fixed in the just-released version (2.0.0)
